### ISSUE TITLE
[Nested Tensor / subclasses] view(-1) and splitting dimensions with view() failures

### ISSUE THREAD

**Issue by davidberard98** [pytorch/pytorch#128649]
## Issue 1: -1 in view

If you have a subclass which is a .view(-1, ...) of a base subclass tensor and you pass it into a torch-compiled function, we fail when trying to create the fake tensor representing the subclass tensor:

Repro

```python
import torch
from torch.nested._internal.nested_tensor import ViewNestedFromBuffer

def get_inputs():
    lengths = torch.randint(2, 1000, (128,), device='cuda')
    offsets = torch.zeros(129, device='cuda', dtype=torch.int32)
    torch.cumsum(lengths, dim=0, out=offsets[1:])

    jagged_input = torch.randn((offsets[-1].item(), 4*32), device='cuda', requires_grad=True)
    return jagged_input, offsets

def get_nt():
    nt = ViewNestedFromBuffer.apply(*get_inputs())
    return nt

def fn(nt):
    return nt.sin()

def fn2(nt):
    return nt.cos()

# Issue 1
torch.compile(fn)(get_nt().view(-1, -1, 128))

# Issue 2
nt = get_nt()
torch.compile(fn2)(nt.view(nt.shape[0], nt.shape[1], 4, 32))
```

Error
```
  File "/home/dberard/local/pytorch/torch/_dynamo/variables/builder.py", line 2299, in <lambda>
    lambda: tx.fake_mode.from_tensor(
  File "/home/dberard/local/pytorch/torch/_subclasses/fake_tensor.py", line 1941, in from_tensor
    return self.fake_tensor_converter.from_real_tensor(
  File "/home/dberard/local/pytorch/torch/_subclasses/fake_tensor.py", line 332, in from_real_tensor
    out = self.meta_converter(
  File "/home/dberard/local/pytorch/torch/_subclasses/meta_utils.py", line 1621, in __call__
    r = self.meta_tensor(
  File "/home/dberard/local/pytorch/torch/_subclasses/meta_utils.py", line 1368, in meta_tensor
    r = view_from_base(base, t)
  File "/home/dberard/local/pytorch/torch/_subclasses/meta_utils.py", line 1038, in view_from_base
    fake_t = t.view_func(base, symint_visitor_fn, tensor_visitor_fn)
  File "/home/dberard/local/pytorch/torch/_subclasses/meta_utils.py", line 967, in symint_visitor_fn
    symbol = shape_env.create_symbol(s, sym_source)
  File "/home/dberard/local/pytorch/torch/fx/experimental/recording.py", line 245, in wrapper
    return fn(*args, **kwargs)
  File "/home/dberard/local/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3457, in create_symbol
    assert not (positive and val < 0), f"positive set for negative value: {val}"
AssertionError: positive set for negative value: -1
```

We can 

## Issue 2: Can't simplify some symints from .view-ed subclasses

Repro: same code as Issue 1 above, but comment out the line marked as "Issue 1"

```
...
  File "/home/dberard/local/pytorch/torch/fx/experimental/symbolic_shapes.py", line 4089, in produce_guards
    issue_guard(guard)
  File "/home/dberard/local/pytorch/torch/fx/experimental/symbolic_shapes.py", line 4053, in issue_guard
    guard_expr = ShapeGuardPrinter(symbol_to_source, source_ref, self.var_to_sources).doprint(expr)
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/printer.py", line 292, in doprint
    return self._str(self._print(expr))
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/str.py", line 778, in _print_Relational
    self._print(expr.rhs))
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/str.py", line 366, in _print_Mul
    a_str = [self.parenthesize(x, prec, strict=False) for x in a]
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/str.py", line 366, in <listcomp>
    a_str = [self.parenthesize(x, prec, strict=False) for x in a]
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/str.py", line 37, in parenthesize
    return self._print(item)
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
  File "/home/dberard/local/pytorch/torch/fx/experimental/symbolic_shapes.py", line 1494, in _print_Symbol
    assert self.symbol_to_source.get(expr), (
AssertionError: s7 (could be from ['<ephemeral: symint_visitor_fn>', '<ephemeral: symint_visitor_fn>']) not in {s3: ["L['nt']._values.size()[0]", "L['nt']._values.size()[0]"], s0: ["L['nt'].size()[1]"], s6: ["L['nt'].size()[2]", "L['nt']._values.size()[1]"], s1: ["L['nt']._values.size()[0]"], s7: []}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665
```

### Versions

main branch ~june 11

**Comment by jbschlosser** [comment]
Issue 2 makes me sad. AFAICT the symbol is not being simplified out because it's part of guard expressions:
* `Eq(s5*s6*s7, 60)` from the view decomp. Solving for `s7` gives `Eq(s7, 60/(s5*s6))`, and then the logic attempts to replace `s7` with `60/(s5*s6)`, but the latter doesn't have `is_integer=True`, so this branch is skipped:
https://github.com/pytorch/pytorch/blob/2e5366fbc04f819c62612e8c56fb786b43c1c67d/torch/fx/experimental/symbolic_shapes.py#L4947-L4951
* `Eq(s7, (15//s6))` doesn't allow for symbol replacement either; it throws here:
https://github.com/pytorch/pytorch/blob/2e5366fbc04f819c62612e8c56fb786b43c1c67d/torch/fx/experimental/symbolic_shapes.py#L4913-L4914

TL;DR: can't do symbol replacement for any of the guards so the ephemeral symbol created during view replay can't be simplified out

**Comment by ezyang** [comment]
Here we know that s5, s6 and s7 divide 60 evenly. Thus, the replacement should be allowed. I would probably try a fix along those lines.

However, I've generally found it is better to express things in terms of s5*s6*s7, instead of trying to replace s7 with 60/(s5*s6), so it would be good to understand why the replacement is preferred as well.

**Comment by StrongerXi** [comment]
Doing old issue scrubbing; issue 1 not longer repros, but issue 2 still does, on main 0c8028e8772. @davidberard98 should we still leave this open?

**Comment by jbschlosser** [comment]
@StrongerXi yes please leave it open :)
